Uchu.World/Objects/Components/Player/ModularBuilderComponent.cs

```diff
@@ -1,250 +1,292 @@
+            await StoreCurrentModel(message.Item);
+        }
 
-            var modelLot = await CreateNewModel(parts);
+        /// <summary>
+        /// Is called when a player picks up an item from the build/ground.
+        /// </summary>
+        public async void MoveAndEquip(ModularBuildMoveAndEquipMessage message)
+        {
+            Logger.Debug($"Associate: {message.Associate}");
 
-            // Don't give back the original model if the user made a new one.
-            CurrentModel = null;
+            var inventory = Player.GetComponent<InventoryManagerComponent>();
+
+            var item = inventory.FindItem(message.Lot, InventoryType.TemporaryModels);
+
+            // This is just fallback and should never happen
+            if (item == null)
+                item = inventory.FindItem(message.Lot);
 
-            // Finish the build.
-            await ConfirmFinish();
+            if (item == null)
+                return;
+
+            var source = item.Inventory.InventoryType;
+            var target = InventoryType.Models;
+            await inventory.MoveItemBetweenInventoriesAsync(item, 1, source, target);
+
+            // We have to find this item again because we moved it
+            item = inventory.FindItem(message.Lot, target);
+            if (item == null)
+                return;
+
+            await item.EquipAsync();
+        }
+
+        /// <summary>
+        /// Finishes the build and assembles a model. This new model is put into the models inventory.
+        /// </summary>
+        public async Task FinishBuild(ModularBuildFinishMessage message)
+        {
+            Logger.Debug($"Associate: {message.Associate}");
+
+            var modelLot = await CreateModel(message.Modules);
+
+            await ExitBuild();
             await this.OnBuildFinished.InvokeAsync(new ModularBuildCompleteEvent
             {
                 model = modelLot,
-                parts = parts
+                parts = message.Modules
             });
         }
-        
+
         /// <summary>
-        /// When the player leaves the build mode either through exiting or being done.
+        /// Quits the build without assembling a models. Puts all used items back into the models inventory.
         /// </summary>
-        public void DoneArranging(DoneArrangingWithItemMessage message)
+        public async void ConfirmExitBuild(BuildExitConfirmationMessage message)
         {
+            Logger.Debug($"Associate: {message.Associate}");
+
+            await RetrieveCurrentModel();
+            await ExitBuild();
         }
 
         /// <summary>
-        /// When the player picks up a part from the ground.
+        /// Stops the build mode and cleans up.
         /// </summary>
-        public async Task Pickup(Lot lot)
+        private async Task ExitBuild()
         {
-            if (!(GameObject is Player player)) return;
+            // Should be removed once PopEquippedItemState is implemented
```

Comment:
It seems you have, in fact, implemented PopEquippedItemState, right?

Reply:
Oh right. I forgot that. Will remove that and test again

Reply:
Code is now removed